expr/shaper.go

```diff
@@ -603,10 +626,44 @@ func castType(zctx *zson.Context, inType, specType zng.Type) (zng.Type, error) {
 			return zctx.LookupTypeArray(out), nil
 		}
 		return zctx.LookupTypeSet(out), nil
-	default:
-		// Non-castable type pair with at least one
-		// (non-record) container: output column is left
-		// unchanged.
-		return inType, nil
 	}
+	if bestUnionSelector(specType, inType) > -1 {
+		return specType, nil
+	}
+	// Non-castable type pair with at least one
+	// (non-record) container: output column is left
+	// unchanged.
+	return inType, nil
+}
+
+// bestUnionSelector tries to return the most specific selector for t in
+// maybeUnion.  It returns -1 if maybeUnion is not a union or contains no type
+// compatible with t.  (Types are compatible if they have the same underlying
+// type.)  If maybeUnion contains t, bestUnionSelector returns its selector.
+// Otherwise, if maybeUnion contains t's underlying type, bestUnionSelector
+// returns its selector.  Finally, bestUnionSelector returns the smallest
+// maybeUnion selector whose type is compatible with t.
```

Comment:
Maybe this would be a little easier if `index`/`indexOf` was used in place of `selector`?

Reply:
The choice of "selector" here stems from an exchange with @mccanne about our use in the code of both "index" and "selector" to denote this entity. I have a preference for a single term, and he has a preference for "selector" (reserving "index" for data structures), so I used "selector" here and hope to make the rest of the code consistent in another PR.